hydra-node/src/Hydra/API/RestServer.hs

```diff
@@ -20,62 +43,105 @@ deriving stock instance IsTx tx => Show (DraftCommitTxResponse tx)
   arbitrary = genericArbitrary
 
   shrink = \case
     DraftCommitTxResponse xs -> DraftCommitTxResponse <$> shrink xs
 
-newtype DraftCommitTxRequest tx = DraftCommitTxRequest
-  { utxos :: UTxOType tx
+-- TODO: This should actually be isomorphic to ScriptWitness of cardano-api,
+-- i.e. we should support also native scripts, other versions of plutus and
+-- witnessing via reference inputs
+data ScriptInfo = ScriptInfo
+  { redeemer :: HashableScriptData
+  , datum :: HashableScriptData
+  , plutusV2Script :: PlutusScript
   }
-  deriving (Generic)
+  deriving stock (Show, Eq, Generic)
+  deriving anyclass (ToJSON, FromJSON)
 
-deriving newtype instance IsTx tx => Eq (DraftCommitTxRequest tx)
-deriving newtype instance IsTx tx => Show (DraftCommitTxRequest tx)
+instance Arbitrary ScriptInfo where
+  arbitrary = genericArbitrary
 
-instance (IsTx tx, ToCBOR tx) => ToJSON (DraftCommitTxRequest tx) where
-  toJSON (DraftCommitTxRequest utxo) =
-    object
-      [ "tag" .= String "DraftCommitTxRequest"
-      , "utxos" .= toJSON utxo
-      ]
+data DraftUTxO = DraftUTxO
+  { txOut :: TxOut CtxUTxO
+  , witness :: Maybe ScriptInfo
+  }
+  deriving stock (Show, Eq, Generic)
 
-instance
-  (IsTx tx, FromCBOR tx) =>
-  FromJSON (DraftCommitTxRequest tx)
-  where
-  parseJSON = withObject "DraftCommitTxRequest" $ \o -> do
-    tag <- o .: "tag"
-    case tag :: Text of
-      "DraftCommitTxRequest" -> do
-        utxos :: (UTxOType tx) <- o .: "utxos"
-        pure $ DraftCommitTxRequest utxos
-      _ -> fail "Expected tag to be DraftCommitTxRequest"
-
-instance Arbitrary (UTxOType tx) => Arbitrary (DraftCommitTxRequest tx) where
+instance ToJSON DraftUTxO where
+  toJSON DraftUTxO{txOut, witness} =
+    case toJSON txOut of
+      Object km ->
+        Object $ km & "witness" `KeyMap.insert` toJSON witness
+      x -> x
+
+instance FromJSON DraftUTxO where
+  parseJSON v = do
+    txOut <- parseJSON v
+    flip (withObject "DraftUTxO") v $ \o -> do
+      witness <- o .: "witness"
+      pure $ DraftUTxO{txOut, witness}
+
+instance Arbitrary DraftUTxO where
+  arbitrary = genericArbitrary
+
+deriving newtype instance Arbitrary (UTxO' DraftUTxO)
+
+newtype DraftCommitTxRequest = DraftCommitTxRequest
+  { utxos :: UTxO' DraftUTxO
+  }
+  deriving stock (Eq, Show, Generic)
+  deriving anyclass (ToJSON, FromJSON)
+
+instance Arbitrary DraftCommitTxRequest where
   arbitrary = genericArbitrary
 
   shrink = \case
-    DraftCommitTxRequest xs -> DraftCommitTxRequest <$> shrink xs
+    DraftCommitTxRequest u -> DraftCommitTxRequest <$> shrink u
+
+convertDraftUTxO ::
+  UTxO' DraftUTxO ->
+  UTxO' (TxOut CtxUTxO, Maybe (ScriptWitness WitCtxTxIn Era))
+convertDraftUTxO utxo' =
+  (\DraftUTxO{txOut, witness} -> (txOut, toScriptWitness <$> witness)) <$> utxo'
+ where
+  toScriptWitness ScriptInfo{redeemer, datum, plutusV2Script} =
+    mkScriptWitness plutusV2Script (ScriptDatumForTxIn datum) redeemer
+
+prepareCommitTxInputs ::
```

Comment:
If we change interfaces as I have outlined in other comments, this should not be needed.